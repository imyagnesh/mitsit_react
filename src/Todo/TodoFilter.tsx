import classNames from "classnames";
import React, { memo } from "react";
import { FilterType } from ".";

type Props = {
  filterTodo: (filterType: FilterType) => void;
  filterType: FilterType;
};

const TodoFilter = ({ filterTodo, filterType }: Props) => {
  // console.log("TodoFilter render");

  return (
    <div className="flex">
      <button
        onClick={() => filterTodo("all")}
        className={classNames("flex-1 py-2 bg-blue-500 text-white", {
          "bg-green-500": filterType === "all",
        })}
        type="button"
      >
        All
      </button>
      <button
        onClick={() => filterTodo("pending")}
        className={classNames("flex-1 py-2 bg-blue-500 text-white", {
          "bg-green-500": filterType === "pending",
        })}
        type="button"
      >
        Pending
      </button>
      <button
        onClick={() => filterTodo("completed")}
        className={classNames("flex-1 py-2 bg-blue-500 text-white", {
          "bg-green-500": filterType === "completed",
        })}
        type="button"
      >
        Completed
      </button>
    </div>
  );
};

export default memo(TodoFilter, () => {
  return true;
});
