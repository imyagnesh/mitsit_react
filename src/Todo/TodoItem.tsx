import React, { memo } from "react";
import cn from "classnames";
import { AppStatus, TodoItemType } from ".";

type Props = {
  todoItem: TodoItemType;
  completeTodo: (todoItem: TodoItemType) => void;
  deleteTodo: (id: number) => void;
  status: AppStatus | undefined;
};

const TodoItem = ({ todoItem, completeTodo, deleteTodo, status }: Props) => {
  // console.log("Todo Item render");

  return (
    <div key={todoItem.id} className="flex items-center m-2">
      <input
        type="checkbox"
        disabled={status?.type === "UPDATE_TODO" && status.state === "LOADING"}
        checked={todoItem.isDone}
        onChange={() => completeTodo(todoItem)}
        className="focus:ring-indigo-500 h-4 w-4 text-indigo-600 border-gray-300 rounded disabled:text-gray-600 disabled:ring-gray-500"
      />
      <p
        className={cn("flex-1 px-4", {
          "line-through": todoItem.isDone,
        })}
      >
        {todoItem.text}
      </p>
      <button
        disabled={status?.type === "DELETE_TODO" && status.state === "LOADING"}
        className="btn disabled:bg-slate-400"
        onClick={() => deleteTodo(todoItem.id)}
      >
        Delete
      </button>
    </div>
  );
};

export default memo(TodoItem);
