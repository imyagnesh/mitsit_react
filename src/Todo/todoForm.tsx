import React, { FormEvent, memo, forwardRef } from "react";
import { AppStatus } from ".";

type Props = {
  onSubmit: (event: FormEvent<HTMLFormElement>) => void;
  status: AppStatus | undefined;
};

export type Ref = HTMLInputElement;

const TodoForm = forwardRef<Ref, Props>(({ onSubmit, status }, ref) => {
  // console.log("Todo Form");

  return (
    <form
      data-testid="todoForm_container"
      className="flex justify-center"
      onSubmit={onSubmit}
    >
      <div className="w-full sm:w-1/2 md:w-1/4">
        <label htmlFor="todo-text" className="hidden">
          First name
        </label>
        <input
          ref={ref}
          type="text"
          name="todo-text"
          id="todo-text"
          className="text-input"
        />
      </div>
      <button
        disabled={status && status.state === "LOADING"}
        type="submit"
        className="btn disabled:bg-slate-500"
      >
        Save
      </button>
    </form>
  );
});

TodoForm.displayName = "TodoForm";

export default memo(TodoForm);
