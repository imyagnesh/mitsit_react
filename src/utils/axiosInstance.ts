import axios from 'axios';

const axiosInstance = axios.create({
  baseURL: 'http://localhost:8080',
  timeout: 5000,
  timeoutErrorMessage: 'Timeout. Please try after sometime',
});

axiosInstance.interceptors.request.use(
  (config) => config,
  (error) => Promise.reject(error),
);

// Add a response interceptor
axiosInstance.interceptors.response.use(
  (response) => response,
  (error) => {
    if (axios.isAxiosError(error)) {
      if (error.response?.data) {
        return Promise.reject(new Error(error.response?.data));
      }
    }

    return Promise.reject(error);
  },
);

export default axiosInstance;
