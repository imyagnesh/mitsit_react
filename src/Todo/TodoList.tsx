import React from "react";
import { AppStatus, FilterType, TodoItemType } from ".";
import TodoItem from "./TodoItem";

type Props = {
  todoList: TodoItemType[];
  completeTodo: (todoItem: TodoItemType) => void;
  deleteTodo: (id: number) => void;
  status: AppStatus[];
};

const TodoList = ({ todoList, completeTodo, deleteTodo, status }: Props) => {
  // console.log("TodoList render");

  return (
    <div className="flex-1">
      {todoList.map((item) => {
        return (
          <TodoItem
            key={item.id}
            todoItem={item}
            completeTodo={completeTodo}
            deleteTodo={deleteTodo}
            status={status.find((x) => x.id === item.id)}
          />
        );
      })}
    </div>
  );
};

export default TodoList;
