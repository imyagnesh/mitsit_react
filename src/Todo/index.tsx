import React, {
  Component,
  FormEvent,
  createRef,
  HtmlHTMLAttributes,
} from "react";
import "./todo.scss";
import TodoForm from "./todoForm";
import TodoFilter from "./TodoFilter";
import TodoList from "./TodoList";

export type TodoItemType = {
  id: number;
  text: string;
  isDone: boolean;
};

type Props = {};

export type FilterType = "all" | "pending" | "completed";

type LoadingAppStatus = {
  state: "LOADING";
  error?: never;
};

type ErrorAppStatus = {
  state: "ERROR";
  error: Error;
};

type BaseAppStatus = LoadingAppStatus | ErrorAppStatus;

type WithoutIdType = "LOAD_TODO" | "ADD_TODO";
type WithIdType = "UPDATE_TODO" | "DELETE_TODO";

type StatusType = WithoutIdType | WithIdType;

type WithoutIdTypes = {
  type: WithoutIdType;
  id?: never;
};

type WithIdTypes = {
  type: WithIdType;
  id: number;
};

type BaseTypeStatus = WithoutIdTypes | WithIdTypes;

export type AppStatus = BaseAppStatus & BaseTypeStatus;

type State = {
  todoList: TodoItemType[];
  filterType: FilterType;
  appStatus: AppStatus[];
};

export default class index extends Component<Props, State> {
  state = {
    appStatus: [] as AppStatus[],
    todoList: [] as TodoItemType[],
    filterType: "all" as FilterType,
  };

  inputRef = createRef<HTMLInputElement>();

  async componentDidMount() {
    this.loadData("all");
  }

  setLoadingStatus = (type: StatusType, id?: number) => {
    this.setState(({ appStatus }) => {
      const newStatus: AppStatus = {
        state: "LOADING",
        type,
        id: id,
      } as BaseTypeStatus & LoadingAppStatus;
      return {
        appStatus: [...appStatus, newStatus],
      };
    });
  };

  setSuccessStatus = (type: StatusType, id?: number) => {
    this.setState(({ appStatus }) => {
      return {
        appStatus: appStatus.filter((x) => !(x.type == type && x.id === id)),
      };
    });
  };

  setErrorStatus = (type: StatusType, error: Error, id?: number) => {
    this.setState(({ appStatus }) => {
      return {
        appStatus: appStatus.map((x) => {
          if (x.type === type && x.id === id) {
            return { ...x, state: "ERROR", error: error as Error };
          }
          return x;
        }),
      };
    });
  };

  loadData = async (filterType: FilterType) => {
    const type: StatusType = "LOAD_TODO";
    try {
      this.setLoadingStatus(type);

      let url = "http://localhost:3000/todoList";
      if (filterType !== "all") {
        url = `${url}?isDone=${filterType === "completed"}`;
      }

      const res = await fetch(url);
      const json = await res.json();
      this.setState({
        todoList: json,
        filterType,
      });
      this.setSuccessStatus(type, undefined);
    } catch (error) {
      this.setErrorStatus(type, error as Error);
    }
  };

  onSubmit = async (event: FormEvent<HTMLFormElement>) => {
    const type: StatusType = "ADD_TODO";
    try {
      event.preventDefault();
      if (this.inputRef.current?.value) {
        this.setLoadingStatus(type);
        const res = await fetch("http://localhost:3000/todoList", {
          method: "POST",
          body: JSON.stringify({
            text: this.inputRef.current?.value || "",
            isDone: false,
          }),
          headers: {
            "Content-Type": "application/json",
            Accept: "application/json",
          },
        });

        const json = await res.json();

        // this.loadData();

        this.setState(
          ({ todoList }) => {
            return {
              todoList: [...todoList, json],
            };
          },
          () => {
            if (this.inputRef.current?.value) {
              this.inputRef.current.value = "";
            }
          }
        );
        this.setSuccessStatus(type);
      }
    } catch (error) {
      this.setErrorStatus(type, error as Error);
    }
  };

  completeTodo = async (todoItem: TodoItemType) => {
    const type: StatusType = "UPDATE_TODO";
    try {
      this.setLoadingStatus(type, todoItem.id);
      const res = await fetch(`http://localhost:3000/todoList/${todoItem.id}`, {
        method: "PUT",
        body: JSON.stringify({ ...todoItem, isDone: !todoItem.isDone }),
        headers: {
          "Content-Type": "application/json",
          Accept: "application/json",
        },
      });

      const json = await res.json();

      this.setState(({ todoList }) => {
        // O(logN)
        const index = todoList.findIndex((x) => x.id === todoItem.id);
        return {
          todoList: [
            ...todoList.slice(0, index),
            json,
            ...todoList.slice(index + 1),
          ],
        };
      });
      this.setSuccessStatus(type, todoItem.id);
    } catch (error) {
      this.setErrorStatus(type, error as Error, todoItem.id);
    }
  };

  deleteTodo = async (id: number) => {
    const type: StatusType = "DELETE_TODO";
    try {
      this.setLoadingStatus(type, id);
      await fetch(`http://localhost:3000/todoList/${id}`, {
        method: "DELETE",
      });

      this.setState(({ todoList }) => {
        const index = todoList.findIndex((x) => x.id === id);
        return {
          todoList: [...todoList.slice(0, index), ...todoList.slice(index + 1)],
        };
      });
      this.setSuccessStatus(type, id);
    } catch (error) {
      this.setErrorStatus(type, error as Error, id);
    }
  };

  render() {
    console.log("App rerender");

    const { todoList, filterType, appStatus } = this.state;

    const loadingStatus = appStatus.find((x) => x.type === "LOAD_TODO");

    if (loadingStatus) {
      if (loadingStatus.state === "LOADING") {
        return <h1>Loading...</h1>;
      }
      if (loadingStatus.state === "ERROR") {
        return <h1>{loadingStatus.error.message}</h1>;
      }
    }

    return (
      <div className="container">
        <h1>Todo App</h1>
        <TodoForm
          onSubmit={this.onSubmit}
          ref={this.inputRef}
          status={appStatus.find((x) => x.type === "ADD_TODO")}
        />
        <TodoList
          todoList={todoList}
          completeTodo={this.completeTodo}
          deleteTodo={this.deleteTodo}
          status={appStatus.filter(
            (x) => x.type === "UPDATE_TODO" || x.type === "DELETE_TODO"
          )}
        />
        <TodoFilter filterType={filterType} filterTodo={this.loadData} />
      </div>
    );
  }
}
