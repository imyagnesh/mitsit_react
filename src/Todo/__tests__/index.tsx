import * as React from "react";
import MockAdapter from "axios-mock-adapter";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import axiosInstance from "../../utils/axiosInstance";
import Todo from "../index";

const mock = new MockAdapter(axiosInstance);

describe("Todo Component", () => {
  test("should render component", () => {
    const { container } = render(<Todo />);
    expect(container.firstChild).toBeInTheDocument();
  });

  test("should display loading on component mount", () => {
    render(<Todo />);
    const pageLoading = screen.queryByTestId("load_todo");
    expect(pageLoading).toBeInTheDocument();
  });

  test("should fail load todo", async () => {
    render(<Todo />);
    const errorLoading = await screen.findByTestId("load_todo_fail");
    expect(errorLoading).toBeInTheDocument();
  });

  test("should display todo container", async () => {
    mock.onGet("todoList").reply(200, [
      {
        text: "get milk",
        isDone: true,
        id: 1,
      },
    ]);
    render(<Todo />);
    const todoContainer = await screen.findByTestId("todo_container");
    expect(todoContainer).toBeInTheDocument();
  });

  test("should display proper error message", async () => {
    mock.onGet("todoList").networkError();
    render(<Todo />);
    const errorLoading = await screen.findByTestId("load_todo_fail");
    expect(errorLoading).toBeInTheDocument();
    expect(errorLoading.innerHTML).toMatch("Network Error");
  });

  test("should display timeout error Message", async () => {
    mock.onGet("todoList").timeout();
    render(<Todo />);
    const errorLoading = await screen.findByTestId("load_todo_fail");
    expect(errorLoading).toBeInTheDocument();
    expect(errorLoading.innerHTML).toMatch(
      "Timeout. Please try after sometime"
    );
  });

  test("should display todo form", async () => {
    mock.onGet("todoList").reply(200, [
      {
        text: "get milk",
        isDone: true,
        id: 1,
      },
    ]);

    mock.onPost("todoList").reply(200, {
      text: "get Food",
      isDone: true,
      id: 1,
    });

    render(<Todo />);
    const todoContainer = await screen.findByTestId("todo_container");
    expect(todoContainer).toBeInTheDocument();
    const todoFormContainer = screen.queryByTestId("todoForm_container");
    expect(todoFormContainer).toBeInTheDocument();
    const todoTextInput = screen.queryByRole("textbox");
    expect(todoTextInput).toBeInTheDocument();
    if (todoTextInput) {
      fireEvent.change(todoTextInput, {
        target: {
          value: "Get Food",
        },
      });
      const submitButton = screen.queryByRole("button", {
        name: "Save",
      });
      if (submitButton) {
        fireEvent.click(submitButton);
        await waitFor(() => expect(submitButton).toBeDisabled());
        await waitFor(() => expect(submitButton).toBeEnabled());
      }
    }
  });
});
