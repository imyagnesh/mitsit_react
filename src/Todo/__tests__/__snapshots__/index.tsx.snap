// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo Component should take snapshot 1`] = `
<h1>
  Loading...
</h1>
`;
